p2panda-rs/src/test_utils/node/node.rs

```diff
@@ -0,0 +1,372 @@
+                            // And add 2
+                            Some(mut id) => {
+                                id += 2;
+                                id
+                            }
+                            // If there aren't any then this is the first log, return 1
+                            None => 1,
+                        }
+                    }
+                },
+                // If there aren't any then this is the first log, return 1
+                None => 1,
+            },
+        };
+        Ok(LogId::new(log_id))
+    }
+
+    /// Get an array of all entries in database
+    pub fn all_entries(&self) -> Vec<LogEntry> {
+        let mut all_entries: Vec<LogEntry> = Vec::new();
+        self.entries.iter().for_each(|(_id, author_logs)| {
+            author_logs
+                .iter()
+                .for_each(|(_id, log)| all_entries.append(log.entries().as_mut()))
+        });
+        all_entries
+    }
+
+    /// Return the entire database
+    pub fn db(&self) -> Database {
+        self.entries.clone()
+    }
+
+    /// Returns the log id, sequence number, skiplink and backlink hash for a given author and
+    /// schema. All of this information is needed to create and sign a new entry.
+    pub fn next_entry_args(&mut self, author: &Author, schema: &Hash) -> Result<NextEntryArgs> {
+        // Find out the log id for the given schema
+        let log_id = self.get_log_id(&schema, &author)?;
+
+        // Find any logs by this author for this schema
+        let author_log = match self.get_author_entries_mut(author) {
+            Some(logs) => {
+                match logs.values().find(|log| log.schema() == schema.as_str()) {
+                    Some(log) => Some(log),
+                    // No logs by this author for this schema
+                    None => None,
+                }
+            }
+            // No logs for this author
+            None => None,
+        };
+
+        // Find out the sequence number, skip- and backlink hash for the next entry in this log
+        let entry_args = match author_log {
+            Some(log) => {
+                // Next sequence number
+                let seq_num = SeqNum::new((log.entries().len() + 1) as i64).unwrap();
+
+                // Calculate backlink and skiplink
+                let (backlink, skiplink) = calculate_links(&seq_num, log);
+
+                NextEntryArgs {
+                    log_id,
+                    seq_num,
+                    skiplink,
+                    backlink,
+                }
+            }
+            // This is the first entry in the log ..
+            None => NextEntryArgs {
+                log_id,
+                seq_num: SeqNum::new(1).unwrap(),
+                skiplink: None,
+                backlink: None,
+            },
+        };
+        Ok(entry_args)
+    }
+
+    /// Calculate the next entry arguments *at a certain point* in this log. This is helpful
```

Comment:
Very nice! Could this be merged with `next_entry_args`, making `seq_num` an optional parameter there?

Reply:
Good, shout, done.